### DISCUSSION TITLE
Dynamic Bank Selection based on selected track

### DISCUSSION THREAD

**Discussion by Ahmed-Abdelgawad** [tttapa/Control-Surface#998] [Q&A]
I have so far enjoyed using this library a lot! It works great and creates a lot of joy in building midi controllers!

I have created a track_select buttons and leds using MCU and its working great. 5 Buttons and 5 track selects. For each track, I would like to be able to send out 5 unique MIDI CC messages, therefore I have used the Bankable and I tested that it works with 2 extra buttons for changing the banks.

My challenge lies in dynamically linking the track selection with the bank assignment. Ideally, when a track is selected (which is currently working by the Buttons and DAW sides), the bank number should automatically update to match the selected track. This ensures that for each track the dedicated set of unique buttons are assigned to it.

```
NoteButton trackSelectButtons[] = {
    {trackSelectPin[0], {MCU::SELECT_1}}, // Button for Track 1 Select
    {trackSelectPin[1], {MCU::SELECT_2}}, // Button for Track 2 Select
    {trackSelectPin[2], {MCU::SELECT_3}}, // Button for Track 3 Select
    {trackSelectPin[3], {MCU::SELECT_4}}, // Button for Track 4 Select
    {trackSelectPin[4], {MCU::SELECT_5}}, // Button for Track 5 Select
};

NoteLED trackSelectLEDs[] = {
    {trackSelectLedPin[0], {MCU::SELECT_1}},
    {trackSelectLedPin[1], {MCU::SELECT_2}},
    {trackSelectLedPin[2], {MCU::SELECT_3}},
    {trackSelectLedPin[3], {MCU::SELECT_4}},
    {trackSelectLedPin[4], {MCU::SELECT_5}},
};

IncrementDecrementSelector<5> selector {
    bank,               // Bank to manage
    {BANK_UP, BANK_DOWN},// Bank Push button pins (increment, decrement)
    Wrap::Wrap,         // Wrap around
};

// Instantiate five bankable buttons
Bankable::CCButton uniqueFunctions[] = {
    {{bank, BankType::CHANGE_CHANNEL}, uniqueFunctionsPins[0], {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}},
    {{bank, BankType::CHANGE_CHANNEL}, uniqueFunctionsPins[1], {MIDI_CC::General_Purpose_Controller_2, CHANNEL_1}},
    {{bank, BankType::CHANGE_CHANNEL}, uniqueFunctionsPins[2], {MIDI_CC::General_Purpose_Controller_3, CHANNEL_1}},
    {{bank, BankType::CHANGE_CHANNEL}, uniqueFunctionsPins[3], {MIDI_CC::General_Purpose_Controller_4, CHANNEL_1}},
    {{bank, BankType::CHANGE_CHANNEL}, uniqueFunctionsPins[4], {MIDI_CC::General_Purpose_Controller_5, CHANNEL_1}},
};
```

I would appreciate any guidance from @tttapa or some users insights on achieving this dynamic bank assignment based on track selection. Is there an existing method within the library that facilitates this? Or could you suggest an approach to implement this feature? Thanks a lot!

**Comment by Ahmed-Abdelgawad** [comment]
I found a workaround by modifying the handleUpdate function in the class NoteCCKPLED, there I'm handling the selected track and I select the equivalent bank accordingly, it's more like a surgical approach but it works.

**Reply by tttapa** [reply]
There are two main approaches: you could use the MIDI input callbacks (see e.g. https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d32/MIDI-Input-Callback_8ino-example.html), or you could create your own custom NoteLED class (see e.g. https://tttapa.github.io/Control-Surface-doc/Doxygen/dd/da5/Custom-Note-LED-Input-Element-Callback_8ino-example.html).
